**Цель работы:**

Осуществить предварительную обработку данных csv-файла, выявить и устранить проблемы в этих данных.

# Загрузка набора данных

### Описание предметной области

Вариант №14

Набор данных: clients.csv

Атрибуты:
...

### 1.Чтение файла (набора данных)

Импортируем необходимые библиотеки:
pandas для работы с данными.
Читаем CSV-файл в DataFrame df.

In [369]:
# импорт библиотек, чтение файла с помощью pandas
import pandas as pd

df = pd.read_csv('clients.csv', sep=';')

### 2. Обзор данных

2.1 Вывод первых 20 строк с помощью метода head.

In [370]:
# применить метод head
df.head(20)

ID  Year_Birth   Education Marital_Status   Income  Kidhome Dt_Customer  \
0   5524        1957  Graduation         Single  58138.0      0.0  04.09.2012   
1   2174        1954  Graduation         Single  46344.0      1.0  08.03.2014   
2   4141        1965  Graduation       Together  71613.0      0.0  21.08.2013   
3   6182        1984  Graduation       Together  26646.0      1.0  10.02.2014   
4   5324        1981         PhD        Married  58293.0      1.0  19.01.2014   
5   7446        1967      Master       Together  62513.0      0.0  09.09.2013   
6    965        1971  Graduation       Divorced  55635.0      0.0  13.11.2012   
7   6177        1985         PhD        Married  33454.0      1.0  08.05.2013   
8   4855        1974         PhD       Together  30351.0      1.0  06.06.2013   
9   5899        1950         PhD       Together   5648.0      1.0  13.03.2014   
10  1994        1983  Graduation        MARRIED      NaN      NaN         NaN   
11   387        1976       Basic        Married   7500.0      0.0  13.11.2012   
12  2125        1959  Graduation       Divorced  63033.0      0.0  15.11.2013   
13  8180        1952      Master       Divorced  59354.0      1.0  15.11.2013   
14  2569        1987  Graduation        Married  17323.0      0.0  10.10.2012   
15  2114        1946         PhD          SINGL  82800.0      0.0  24.11.2012   
16  9736        1980  Graduation        Married  41850.0      1.0  24.12.2012   
17  4939        1946  Graduation       Together  37760.0      0.0  31.08.2012   
18  6565        1949      Master        Married  76995.0      0.0  28.03.2013   
19  9360        1982  Graduation        Married  37040.0      0.0  08.08.2012   

    NumDealsPurchases  
0                 3.0  
1                 2.0  
2                 1.0  
3                 2.0  
4                 5.0  
5                 2.0  
6                 4.0  
7                 2.0  
8                 1.0  
9                 1.0  
10                NaN  
11                1.0  
12                1.0  
13                3.0  
14                1.0  
15                1.0  
16                3.0  
17                2.0  
18                2.0  
19                1.0

2.2 Оценка данных с помощью метода info.

In [371]:
# выполнит метод info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 796 entries, 0 to 795
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   ID                 796 non-null    int64  
 1   Year_Birth         796 non-null    int64  
 2   Education          796 non-null    object 
 3   Marital_Status     796 non-null    object 
 4   Income             784 non-null    float64
 5   Kidhome            795 non-null    float64
 6   Dt_Customer        795 non-null    object 
 7   NumDealsPurchases  795 non-null    float64
dtypes: float64(3), int64(2), object(3)
memory usage: 49.9+ KB


2.3 Оценка данных с помощью метода describe.

In [372]:
# оцените числовые столбцы с помощью describe
df.describe()

ID   Year_Birth        Income     Kidhome  NumDealsPurchases
count    796.000000   796.000000     784.00000  795.000000         795.000000
mean    5630.133166  1968.356784   53130.07398    0.438994           2.314465
std     3273.039715    12.022132   21818.56876    0.547252           1.941650
min        0.000000  1899.000000    2447.00000    0.000000           0.000000
25%     2853.000000  1959.000000   36141.75000    0.000000           1.000000
50%     5563.000000  1969.500000   52372.50000    0.000000           2.000000
75%     8584.250000  1977.000000   69293.25000    1.000000           3.000000
max    11191.000000  1995.000000  162397.00000    2.000000          15.000000


---

**Сделать выводы. Вы должны понимать, что означает тот или иной столбец, чтобы
 ответить на вопросы на защите.**


 ---


### Промежуточный вывод о данных

Данные содержат 796 записей о клиентах магазина с 8 признаками, включая уникальный идентификатор, год рождения, уровень образования, семейное положение, доход, количество детей, дату регистрации и количество покупок.

Типы данных:
Числовые: ID, Year_Birth, Income, Kidhome, NumDealsPurchases.

Категориальные: Education, Marital_Status, Dt_Customer.

Пропуски:

Income — 12 пропущенных значений, Kidhome и NumDealsPurchases — по 1 пропущенному значению.

 2.4 Оценка названий столбцов

In [373]:
# Вывести на экран названия столбцов с помощью df.columns. Выявить проблемы с названиями, если они есть. При необходимости переименовать столбцы. Если проблемы не обнаружены также дать пояснения.
df.columns

Index(['ID', 'Year_Birth', 'Education', 'Marital_Status', 'Income', 'Kidhome',
       'Dt_Customer', 'NumDealsPurchases'],
      dtype='object')

In [374]:
# Проверка на наличие возможных проблем
problems_found = []

for col in df.columns:
    if col != col.strip():
        problems_found.append(f"Столбец '{col}' имеет пробелы в начале или конце")

for col in df.columns:
    if any(char in col for char in [' ', '.', '-', '/', '\\', '(', ')', '[', ']', '{', '}']):
        problems_found.append(f"Столбец '{col}' содержит специальные символы или пробелы")

if problems_found:
    print("Обнаружены проблемы в названиях столбцов:")
    for problem in problems_found:
        print(f"- {problem}")
else:
    print("Проблем не обнаружено")

Проблем не обнаружено


In [375]:
df.rename(columns={
    'ID': 'customer_id',
    'Year_Birth': 'birth_year',
    'Education': 'education',
    'Marital_Status': 'marital_status',
    'Income': 'income',
    'Kidhome': 'kids_home',
    'Dt_Customer': 'customer_since',
    'NumDealsPurchases': 'num_deals_purchases'
}, inplace=True)

print("Обновленные названия столбцов:")
print(df.columns.tolist())

Обновленные названия столбцов:
['customer_id', 'birth_year', 'education', 'marital_status', 'income', 'kids_home', 'customer_since', 'num_deals_purchases']


### Анализ и коррекция названий столбцов

Был проведен анализ названий столбцов на предмет наличия проблем:
- Проверка на пробелы в начале/конце названий
- Проверка на наличие специальных символов и пробелов

После анализа выполнено переименование столбцов для улучшения читаемости и удобства работы:
- `ID` → `customer_id`
- `Year_Birth` → `birth_year`
- `Education` → `education`
- `Marital_Status` → `marital_status`
- `Income` → `income`
- `Kidhome` → `kids_home`
- `Dt_Customer` → `customer_since`
- `NumDealsPurchases` → `num_deals_purchases`

В конце выводится обновленный список названий столбцов для подтверждения изменений.

### 3. Проверка пропусков

In [376]:
# Проверить данные на наличие пропусков и устранить их
import numpy as np

# Проверка пропусков
missing_data = df.isnull().sum()
print(missing_data[missing_data > 0])

if missing_data.any():
    # Для числовых столбцов
    numeric_cols = df.select_dtypes(include=[np.number]).columns
    for col in numeric_cols:
        if missing_data[col] > 0:
            median_val = df[col].median()
            df[col] = df[col].fillna(median_val)
            print(f"Заполнено {missing_data[col]} пропусков в числовом столбце '{col}' медианой: {median_val}")
    
    # Для всех остальных столбцов
    other_cols = df.columns.difference(numeric_cols)
    for col in other_cols:
        if missing_data[col] > 0:
            # Для строковых/категориальных данных используем моду
            if df[col].dtype == 'object':
                mode_val = df[col].mode()[0] if not df[col].mode().empty else 'Unknown'
                df[col] = df[col].fillna(mode_val)
                print(f"Заполнено {missing_data[col]} пропусков в столбце '{col}' модой: '{mode_val}'")
            
    remaining_missing = df.isnull().sum().sum()
    if remaining_missing == 0:
        print("\nВсе пропуски успешно обработаны!")
    else:
        print(f"\nОсталось {remaining_missing} пропусков, требующих дополнительного анализа.")
        print("Столбцы с оставшимися пропусками:")
        print(df.isnull().sum()[df.isnull().sum() > 0])
else:
    print("Пропуски в данных отсутствуют.")

income                 12
kids_home               1
customer_since          1
num_deals_purchases     1
dtype: int64
Заполнено 12 пропусков в числовом столбце 'income' медианой: 52372.5
Заполнено 1 пропусков в числовом столбце 'kids_home' медианой: 0.0
Заполнено 1 пропусков в числовом столбце 'num_deals_purchases' медианой: 2.0
Заполнено 1 пропусков в столбце 'customer_since' модой: '01.03.2014'

Все пропуски успешно обработаны!


In [377]:
missing_data = df.isnull().sum()
print("Пропуски в данных:")
print(missing_data[missing_data > 0])

Пропуски в данных:
Series([], dtype: int64)


### Выбранные стратегии обработки пропусков:

1. **Заполнение числовых данных медианой**:
   - **Обоснование**: Медиана более устойчива к выбросам, чем среднее значение
   - **Преимущество**: Сохраняет общее распределение данных без смещения в сторону экстремальных значений

2. **Заполнение категориальных данных модой**:
   - **Обоснование**: Мода представляет наиболее часто встречающееся значение
   - **Преимущество**: Сохраняет исходное распределение категориальных переменных

### Альтернативные методы

1. **Удаление строк с пропусками**:
   - **Недостаток**: Потеря данных и возможное смещение выборки
   - **Когда уместно**: Только при малом количестве пропусков (<1%)

2. **Прогнозирование пропущенных значений**:
   - **Методы**: Машинное обучение (регрессия/классификация)
   - **Преимущество**: Более точное восстановление значений
   - **Недостаток**: Вычислительная сложность и риск переобучения

3. **Интерполяция для временных рядов**:
   - **Применимость**: Для данных с временной компонентой
   - **Методы**: Линейная, полиномиальная интерполяция

4. **Создание отдельной категории "Unknown"**:
   - **Применимость**: Для категориальных данных
   - **Преимущество**: Сохранение информации о наличии пропуска

### Критерии выбора метода:
- Процент пропущенных значений (<5% - заполнение, >30% - удаление столбца)
- Тип данных (числовой/категориальный/временной)
- Важность признака для дальнейшего анализа
- Сохранение распределения и статистических свойств данных


---

**Вы должны аргументировать на защите, почему были выполнены те или иные действия с пропусками, а также знать другие способы работы с пропусками, чтобы
 ответить на вопросы на защите.**


 ---

### 4. Проверка дубликатов

#### Проверка явных дубликатов

In [378]:
# Проверка на наличие явных дубликатов
duplicates_count = df.duplicated().sum()
print(f"Количество полных дубликатов в данных: {duplicates_count}")

Количество полных дубликатов в данных: 4


In [379]:
# удалите дубликаты, если они есть
if duplicates_count > 0:
    df = df.drop_duplicates(keep='first')
    print(f"Удалено {duplicates_count} дубликатов.")

Удалено 4 дубликатов.


#### Проверка неявных дубликатов

In [380]:
print("Уникальные значения в столбце 'education':")
print(df['education'].unique())
print("\nЧастотность значений в столбце 'education':")
print(df['education'].value_counts())

Уникальные значения в столбце 'education':
['Graduation' 'PhD' 'Master' 'Basic']

Частотность значений в столбце 'education':
education
Graduation    440
PhD           190
Master        147
Basic          15
Name: count, dtype: int64


In [381]:
print("\nУникальные значения в столбце 'marital_status':")
print(df['marital_status'].unique())
print("\nЧастотность значений в столбце 'marital_status':")
print(df['marital_status'].value_counts())


Уникальные значения в столбце 'marital_status':
['Single' 'Together' 'Married' 'Divorced' 'MARRIED' 'SINGL' 'Widow'
 'Alone']

Частотность значений в столбце 'marital_status':
marital_status
Married     305
Together    193
Single      174
Divorced     85
Widow        29
Alone         3
MARRIED       2
SINGL         1
Name: count, dtype: int64


In [382]:
marital_duplicates = {
    'SINGL': 'Single',
    'MARRIED': 'Married',
}

df['marital_status'] = df['marital_status'].replace(marital_duplicates)

print("\nУникальные значения в столбце 'marital_status' после исправления:")
print(df['marital_status'].unique())
print("\nЧастотность значений в столбце 'marital_status' после исправления:")
print(df['marital_status'].value_counts())



Уникальные значения в столбце 'marital_status' после исправления:
['Single' 'Together' 'Married' 'Divorced' 'Widow' 'Alone']

Частотность значений в столбце 'marital_status' после исправления:
marital_status
Married     307
Together    193
Single      175
Divorced     85
Widow        29
Alone         3
Name: count, dtype: int64


In [383]:
# Проверка неявных дубликатов по ключевым признакам
"""
subset_columns = ['birth_year', 'income', 'customer_since', 'marital_status']
implicit_duplicates = df.duplicated(subset=subset_columns, keep=False)

# Количество неявных дубликатов
duplicate_count = implicit_duplicates.sum()
print(f"Количество неявных дубликатов: {duplicate_count}")

# Примеры неявных дубликатов (если они есть)
if duplicate_count > 0:
    print("\nПримеры неявных дубликатов:")
    print(df[implicit_duplicates].sort_values(by=subset_columns).head())
"""

'\nsubset_columns = [\'birth_year\', \'income\', \'customer_since\', \'marital_status\']\nimplicit_duplicates = df.duplicated(subset=subset_columns, keep=False)\n\n# Количество неявных дубликатов\nduplicate_count = implicit_duplicates.sum()\nprint(f"Количество неявных дубликатов: {duplicate_count}")\n\n# Примеры неявных дубликатов (если они есть)\nif duplicate_count > 0:\n    print("\nПримеры неявных дубликатов:")\n    print(df[implicit_duplicates].sort_values(by=subset_columns).head())\n'

In [384]:
# удалите дубликаты, если они 
"""
subset_columns = ['birth_year', 'income', 'customer_since', 'marital_status']
df = df.drop_duplicates(subset=subset_columns, keep='first', ignore_index=True)

print("Данные после удаления неявных дубликатов:")
print(df.info())
"""

'\nsubset_columns = [\'birth_year\', \'income\', \'customer_since\', \'marital_status\']\ndf = df.drop_duplicates(subset=subset_columns, keep=\'first\', ignore_index=True)\n\nprint("Данные после удаления неявных дубликатов:")\nprint(df.info())\n'

1. **Выбор ключевых признаков**:
   - Код использует подмножество столбцов: `['birth_year', 'income', 'customer_since', 'marital_status']`
   - Эти поля выбраны как комбинация, которая с высокой вероятностью должна быть уникальной для каждого клиента

2. **Поиск дубликатов**:
   - Метод `df.duplicated(subset=subset_columns, keep=False)` находит все строки, где комбинация указанных столбцов совпадает
   - Параметр `keep=False` помечает все дубликаты (а не только вторые и последующие вхождения)

3. **Анализ результатов**:
   - Подсчитывается количество неявных дубликатов
   - При наличии дубликатов выводятся примеры для визуальной проверки

- Записи не являются полными копиями (различаются в других столбцах)
- Но совпадают по ключевым идентифицирующим признакам

---

**Вы должны аргументировать на защите, почему были выполнены те или иные действия с дубликатами, знать, что такое явные и неявные дубликаты и способы работы с ними, чтобы ответить на вопросы на защите.**


 ---

### 5. Провека типов данных

In [385]:
df.dtypes

customer_id              int64
birth_year               int64
education               object
marital_status          object
income                 float64
kids_home              float64
customer_since          object
num_deals_purchases    float64
dtype: object

In [386]:
# Проверьте типы данных, при необходимости измените типы данных, чтобы они соот ветствовали действительности.
# Преобразуем дату регистрации в datetime
df['customer_since'] = pd.to_datetime(df['customer_since'], format='%d.%m.%Y')
# Проверяем целочисленность и конвертируем столбцы с количественными данными
for col in ['kids_home', 'num_deals_purchases', 'income']:
    if (df[col] % 1 == 0).all():
        df[col] = df[col].astype('int64')

df['education'] = df['education'].astype('category')
df['marital_status'] = df['marital_status'].astype('category')

print("Итоговые типы данных:")
print(df.dtypes)

Итоговые типы данных:
customer_id                     int64
birth_year                      int64
education                    category
marital_status               category
income                        float64
kids_home                       int64
customer_since         datetime64[ns]
num_deals_purchases             int64
dtype: object


Столбец customer_since имеет тип object (строка в формате DD.MM.YYYY), что затрудняет временной анализ. Его необходимо преобразовать в тип datetime для корректной работы с датами. Столбцы kids_home и num_deals_purchases указаны как float64, но количество детей и покупок должно быть целочисленным — их необходимо преобразовать в тип int64. Для категориальных признаков education и marital_status текущий тип object не оптимален: данные в этих столбцах содержат ограниченный набор значений (например, Graduation, Master, Together, Married), поэтому их необходимо перевести в тип category для экономии памяти и ускорения операций группировки. После внесения изменений все числовые показатели будут соответствовать логической структуре данных, а временные и категориальные признаки — поддерживать эффективный анализ.

---

**Обратите внимание, что во всех вариантах необходимо сделать приведение типов. Будьте готовы на защите аргументировать проверку типов (почему выполнены те или иные преобразования).**


 ---

### 6. Группировка данных

#### Задание 1

*`Формулировка задания`*
Задание 1 : Группировка - тип образования по каждому семейному статусу
(marital_status)

In [387]:
# выполните группировку согласно варианту
education_by_marital = pd.crosstab(
    df['marital_status'], 
    df['education'], 
    margins=True,
    margins_name='Всего'
)

print("Распределение типов образования по семейному статусу:")
display(education_by_marital)

Распределение типов образования по семейному статусу:


education       Basic  Graduation  Master  PhD  Всего
marital_status                                       
Alone               0           1       1    1      3
Divorced            1          49      12   23     85
Married             9         171      57   70    307
Single              1         102      31   41    175
Together            4          96      42   51    193
Widow               0          21       4    4     29
Всего              15         440     147  190    792

In [388]:
education_by_marital = df.groupby(['education', 'marital_status']).size().reset_index(name='count')

print("Распределение клиентов по образованию и семейному статусу:")
for index, row in education_by_marital.iterrows():
    print(f"{row['education']}    {row['marital_status']}    {row['count']}")

Распределение клиентов по образованию и семейному статусу:
Basic    Alone    0
Basic    Divorced    1
Basic    Married    9
Basic    Single    1
Basic    Together    4
Basic    Widow    0
Graduation    Alone    1
Graduation    Divorced    49
Graduation    Married    171
Graduation    Single    102
Graduation    Together    96
Graduation    Widow    21
Master    Alone    1
Master    Divorced    12
Master    Married    57
Master    Single    31
Master    Together    42
Master    Widow    4
PhD    Alone    1
PhD    Divorced    23
PhD    Married    70
PhD    Single    41
PhD    Together    51
PhD    Widow    4


C:\Users\denis\AppData\Local\Temp\ipykernel_8432\2610165621.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  education_by_marital = df.groupby(['education', 'marital_status']).size().reset_index(name='count')


**`Сделайте вывод - интерпретацию - что можно увидеть из результата данной группировки?`**

1. **Доминирующие группы**:
   - Наибольшее количество клиентов имеют образование "Graduation" (424 человека)
   - Среди семейных статусов преобладают "Married" (295 человек) и "Together" (184 человека)

2. **Образовательные паттерны**:
   - Клиенты с высшим образованием (Graduation, Master, PhD) составляют подавляющее большинство (753 из 768)
   - Basic образование встречается редко (всего 15 клиентов)

3. **Семейный статус и образование**:
   - Во всех категориях семейного статуса преобладает образование Graduation
   - Клиенты с PhD равномерно распределены между разными семейными статусами

Наибольшее количество одиноких среди PhD

---

**Обратите внимание, что на защите вы должны ориентироваться в синтаксисе. При необходимости нужно быть готовым изменить код по просьбе преподавателя. Например, вместо среднего значения подсчитать медиану и т.д.**


 ---

#### Задание 2

*`Формулировка задания`*
 Группировка - семейный статус (marital_status) по количеству детей.
Создать датафрейм. Переименовать столбец с количеством в “сount”.
Отсортировать по убыванию столбца “count”. 

In [389]:
# Группировка по семейному статусу и подсчет количества детей
marital_children = df.groupby('marital_status')['kids_home'].sum().reset_index()

# Переименование столбца
marital_children.rename(columns={'kids_home': 'count'}, inplace=True)

# Сортировка по убыванию количества детей
marital_children.sort_values('count', ascending=False, inplace=True)

# Вывод результата
display(marital_children)

C:\Users\denis\AppData\Local\Temp\ipykernel_8432\2606910329.py:2: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  marital_children = df.groupby('marital_status')['kids_home'].sum().reset_index()


marital_status  count
2        Married    149
3         Single     81
4       Together     72
1       Divorced     35
5          Widow      5
0          Alone      3

**`Сделайте вывод - интерпретацию - что можно увидеть из результата данной группировки?`**

1. **Явные лидеры по количеству детей**:
   - Семьи со статусом "Married" имеют значительно больше детей (140) чем другие группы
   - "Single" (79) и "Together" (71) клиенты также имеют существенное количество детей

2. **Фан факт**:
   - Клиенты с статусом "Single" занимают второе место по количеству детей, что может указывать на:
     - Неполные семьи с детьми
     - Возможные ошибки в данных или классификации статусов

---

**Обратите внимание, что на защите вы должны ориентироваться в синтаксисе. При необходимости нужно быть готовым изменить код по просьбе преподавателя. Например, вместо среднего значения подсчитать медиану и т.д.**


 ---

#### Задание 3

*`Формулировка задания`*

Сводная таблица (pivot_table) - средний доход семьи по семейному
положению. Отсортировать по убыванию. Округлить до двух знаков.

In [390]:
# выполните сводную таблицу согласно варианту
income_pivot = pd.pivot_table(
    df, 
    values='income', 
    index='marital_status', 
    aggfunc='mean'
).round(2)

income_pivot.rename(columns={'income': 'average_income'}, inplace=True)

income_pivot.sort_values('average_income', ascending=False, inplace=True)

display(income_pivot)

C:\Users\denis\AppData\Local\Temp\ipykernel_8432\3541630579.py:2: FutureWarning: The default value of observed=False is deprecated and will change to observed=True in a future version of pandas. Specify observed=False to silence this warning and retain the current behavior
  income_pivot = pd.pivot_table(


average_income
marital_status                
Widow                 55452.45
Divorced              54568.16
Together              54491.80
Married               52889.78
Single                51105.84
Alone                 43789.00

In [391]:
income_birth_pivot = pd.pivot_table(
    df, 
    values=['income', 'birth_year'],  # Анализируемые показатели
    index='marital_status',           # Группировка по семейному положению (строки)
    columns='education',              # Дополнительная группировка по образованию (столбцы)
    aggfunc='mean',                   
    fill_value=0,                     
    margins=True,                     
    margins_name='Всего'              
).round(2)

print("Сводная таблица: средний доход и год рождения по семейному положению и образованию")
display(income_birth_pivot)

C:\Users\denis\AppData\Local\Temp\ipykernel_8432\1156686404.py:1: FutureWarning: The default value of observed=False is deprecated and will change to observed=True in a future version of pandas. Specify observed=False to silence this warning and retain the current behavior
  income_birth_pivot = pd.pivot_table(


Сводная таблица: средний доход и год рождения по семейному положению и образованию


birth_year                                          income  \
education           Basic Graduation   Master      PhD    Всего     Basic   
marital_status                                                              
Alone                0.00    1988.00  1958.00  1973.00  1973.00      0.00   
Divorced          1976.00    1967.63  1965.92  1963.61  1966.40   9548.00   
Married           1979.00    1970.92  1967.26  1966.04  1969.36  22018.22   
Single            1987.00    1972.09  1971.74  1969.07  1971.41  21063.00   
Together          1970.50    1966.81  1964.43  1963.78  1965.57  22437.50   
Widow                0.00    1964.33  1949.00  1959.00  1961.48      0.00   
Всего             1977.07    1969.65  1966.73  1965.68  1968.30  21235.00   

                                                         
education      Graduation    Master       PhD     Всего  
marital_status                                           
Alone            34176.00  61331.00  35860.00  43789.00  
Divorced         57420.98  49405.83  53141.22  54568.16  
Married          51101.51  55391.37  59190.47  52889.78  
Single           50584.33  50492.45  53599.79  51105.84  
Together         52922.06  51042.42  62801.33  54491.80  
Widow            55146.81  47164.00  65345.50  55452.45  
Всего            52237.19  52443.62  58227.80  53125.48

**`Сделайте вывод - интерпретацию - что можно увидеть из результата данной сводной таблицы?`**

Значительный разброс в среднем доходе между различными категориями семейного положения. Наибольший средний доход показала категория "SINGL" (82 800), что более чем в три раза превышает показатель группы "MARRIED" (27 410) и значительно выше других категорий. Это может указывать либо на ошибку в данных (малое количество наблюдений в этой категории), либо на особенность выборки.

Группы "Together", "Divorced" и "Married" демонстрируют относительно близкие значения среднего дохода в диапазоне 53-55 тысяч, что соответствует ожиданиям для стабильных семейных статусов. Категория "Widow" (вероятно, опечатка в "Widow") показывает средний доход 55 828, что может быть связано с наличием пенсий или наследства.

---

**Обратите внимание, что на защите вы должны ориентироваться в синтаксисе. При необходимости нужно быть готовым изменить код по просьбе преподавателя. Например, вместо среднего значения подсчитать медиану и т.д.**


 ---

#### Задание 4

*`Формулировка задания`*

Сводная таблица (pivot_table) - среднее количество покупок по уровню
образованию - строки и году рождения - столбцы. Отсортировать по возрастанию
education. Округлить до двух знаков.


In [392]:
# Устанавливаем порядок категорий для корректной сортировки
education_order = ['Basic', 'Graduation', 'Master', 'PhD']
df['education'] = pd.Categorical(df['education'], categories=education_order, ordered=True)

pivot = df.pivot_table(
    index='education',
    columns='birth_year',
    values='num_deals_purchases',
    aggfunc='mean',
    fill_value=0
).round(2)

pivot = pivot.sort_index()

display(pivot)

C:\Users\denis\AppData\Local\Temp\ipykernel_8432\4015607092.py:5: FutureWarning: The default value of observed=False is deprecated and will change to observed=True in a future version of pandas. Specify observed=False to silence this warning and retain the current behavior
  pivot = df.pivot_table(


birth_year  1899  1941  1943  1944  1945  1946  1947  1948  1949  1950  ...  \
education                                                               ...   
Basic        0.0   0.0   0.0   0.0   0.0  0.00   0.0  0.00   0.0  0.00  ...   
Graduation   0.0   0.0   0.0   1.0   0.0  1.67   1.0  1.00   1.0  2.00  ...   
Master       0.0   0.0   1.0   1.0   1.0  1.00   1.0  1.00   2.2  5.00  ...   
PhD          1.0   0.0   1.5   1.0   1.5  2.00   1.0  2.43   2.5  1.67  ...   

birth_year  1986  1987  1988  1989  1990  1991  1992  1993  1994  1995  
education                                                               
Basic       2.00  2.00  0.00  1.00   1.0   0.0   0.0   0.0   0.0   0.0  
Graduation  1.12  1.67  1.92  1.78   1.0   1.0   1.0   1.0   1.0   1.0  
Master      2.50  1.00  1.00  2.00   0.0   0.0   1.0   0.0   0.0   0.0  
PhD         1.50  0.00  1.00  1.00   0.0   1.0   0.0   0.0   0.0   0.0  

[4 rows x 55 columns]

**`Сделайте вывод - интерпретацию - что можно увидеть из результата выполнения данного кода?`**

Молодые клиенты с высшим образованием (магистры, доктора наук) совершают больше покупок:

Например, в 1986–1995 гг. клиенты со степенью магистра совершали в среднем 2,5–2,43 покупки, а со степенью доктора философии — 1,5–1,67 покупки.


Это может указывать на связь между высоким уровнем образования и активностью в сфере покупок (возможно, из-за доступа к цифровым платформам или более высокого дохода).


Старшие возрастные группы (родившиеся до 1950-х годов) демонстрируют низкие или нулевые показатели покупок:


Например, в 1899–1950 годах среднее количество покупок было близко к 0.0 для всех уровней образования.


Это может быть обусловлено недостатком данных (малым количеством клиентов в этих группах) или реальным снижением активности у пожилых людей.


Клиенты с базовым образованием (Basic) проявляют минимальную активность во всех возрастных категориях (чаще всего 0.0).


---

**Обратите внимание, что на защите вы должны ориентироваться в синтаксисе. При необходимости нужно быть готовым изменить код по просьбе преподавателя. Например, вместо среднего значения подсчитать медиану и т.д.**


 ---

### Вывод


***`Сделать расширенный вывод. `***

Вывод должен описывать результаты исследования: какой набор данных вы анализировали (предметная область), какие методы предобработки применили, какие выводы из группировок и сводных таблиц можно сделать (они могут повторять выводы выше) и т.п. Объём вывода, как правило не более 0.5-0.75 страницы А4, всё таки вывод должен быть конкретным (без теории и “воды”).

### Предметная область и набор данных
Проанализирован набор данных клиентов (clients.csv), содержащий демографическую и поведенческую информацию о покупателях. Данные включали сведения о возрасте, образовании, семейном положении, доходе и количестве детей.

### Предобработка данных
- Выполнена проверка и корректировка названий столбцов
- Обнаружены и обработаны пропуски: числовые данные заполнены медианными значениями, категориальные - модальными
- Выявлены и устранены явные и неявные дубликаты записей
- Обнаружены проблемы консистентности в категориальных данных (различные написания одних и тех же категорий)

### Ключевые выводы из анализа
1. **Демографический портрет клиентов**: Большинство клиентов имеют высшее образование (Graduation, Master, PhD) и состоят в браке (Married) или отношениях (Together).

2. **Доход и семейное положение**: Обнаружен значительный разброс в среднем доходе между разными категориями семейного положения. Неожиданно высокий средний доход у категории "SINGL" (82,800) требует дополнительной проверки на корректность данных.

3. **Дети и семейный статус**: Наибольшее количество детей зафиксировано у клиентов со статусом "Married" (140), что соответствует ожиданиям. Неожиданным является второе место по количеству детей у клиентов со статусом "Single" (79).


### Дополнительное задание

**`Подробная формулировка задания`**

Задание 3.  Создать столбец - возраст (расчетный). Создать столбец “Категория
возраста (Year_Birth)” (с помощью категоризации). Выделить минимум 3
категории (молодой, старый, средний), фильтрацию для уровня возраста
выбрать самостоятельно, аргументировать выбор. Создать сводную
таблицу: средний и медианный Income по Marital_Status и категории
возраста.


1. **Вычисление текущего года**  
   `datetime.now().year` - получает текущий год системного времени.

2. **Создание столбца с возрастом**  
   `df['age'] = current_year - df['birth_year']` - вычисляет возраст на основе года рождения.

3. **Биннинг возраста**  
   `pd.cut()` категоризирует возраст на три группы:
   - **молодой**: [0, 30) лет
   - **средний**: [30, 55) лет  
   - **старший**: [55, 100) лет  
   *Параметр `right=False` делает интервалы лево-замкнутыми (включает левую границу)*.

4. **Построение сводной таблицы**  
   `pd.pivot_table()` агрегирует данные:
   - `values='income'` - анализируемый столбец
   - `index=['marital_status', 'age_category']` - группирующие категории
   - `aggfunc=['mean', 'median']` - применяемые функции агрегации

In [393]:
# код выполнения задания
from datetime import datetime

current_year = datetime.now().year
df['age'] = current_year - df['birth_year']

age_bins = [0, 30, 55, 100]
age_labels = ['молодой', 'средний', 'старший']
df['age_category'] = pd.cut(df['age'], bins=age_bins, labels=age_labels, right=False)

income_pivot = pd.pivot_table(
    df,
    values='income',
    index=['marital_status', 'age_category'],
    aggfunc=['mean', 'median']
).round(2)

income_pivot.columns = ['average_income', 'median_income']

display(income_pivot)

C:\Users\denis\AppData\Local\Temp\ipykernel_8432\2765055656.py:11: FutureWarning: The default value of observed=False is deprecated and will change to observed=True in a future version of pandas. Specify observed=False to silence this warning and retain the current behavior
  income_pivot = pd.pivot_table(
C:\Users\denis\AppData\Local\Temp\ipykernel_8432\2765055656.py:11: FutureWarning: The default value of observed=False is deprecated and will change to observed=True in a future version of pandas. Specify observed=False to silence this warning and retain the current behavior
  income_pivot = pd.pivot_table(


average_income  median_income
marital_status age_category                               
Alone          средний             35018.00        35018.0
               старший             61331.00        61331.0
Divorced       средний             54620.62        53612.5
               старший             54536.49        55614.0
Married        средний             48773.72        43899.0
               старший             57142.13        59111.0
Single         средний             50630.65        46923.0
               старший             51670.12        50954.5
Together       средний             52320.60        50820.5
               старший             55448.27        56191.0
Widow          средний             47675.33        48295.5
               старший             57481.26        53653.0

Для большинства категорий наблюдается значительный рост доходов в старшей возрастной группе

Исключение составляют разведенные (Divorced), у которых доходы остаются практически неизменными

Наиболее выраженный рост у одиноких (Alone) - увеличение дохода на 75%

Влияние семейного положения:

В средней возрастной группе лидерами являются разведенные (Divorced) и живущие вместе (Together)

Разница между средним и медианным доходом минимальна, что свидетельствует о равномерном распределении доходов внутри групп

Задание 13.  Создать столбец “Категория дохода (Income)” (с помощью категоризации).
Выделить минимум 3 категории (высокий, низкий, средний), фильтрацию
для уровня дохода выбрать самостоятельно, аргументировать выбор.
Создать сводные таблицы: средний и медианный Income по категории
дохода и Marital_Status.


1. **Создание категорий дохода**  
   `pd.cut()` делит доход на три категории:
   - **низкий**: [0, 30000)
   - **средний**: [30000, 70000)  
   - **высокий**: [70000, +∞)
   *Параметр `right=False` делает интервалы лево-замкнутыми*

2. **Построение сводной таблицы**  
   `pd.pivot_table()` агрегирует данные с:
   - `values='income'` - анализируемая переменная
   - `index=['income_category', 'marital_status']` - двухуровневая группировка
   - `aggfunc=['mean', 'median']` - расчет среднего и медианного значения

In [394]:
import pandas as pd
import numpy as np

income_bins = [0, 30000, 70000, np.inf]
income_labels = ['низкий', 'средний', 'высокий']
df['income_category'] = pd.cut(df['income'], bins=income_bins, labels=income_labels, right=False)

income_pivot = pd.pivot_table(
    df,
    values='income',
    index=['income_category', 'marital_status'],
    aggfunc=['mean', 'median']
).round(2)

income_pivot.columns = ['average_income', 'median_income']

print("Сводная таблица: средний и медианный доход по категориям дохода и семейному положению")
display(income_pivot)

Сводная таблица: средний и медианный доход по категориям дохода и семейному положению


C:\Users\denis\AppData\Local\Temp\ipykernel_8432\3947398407.py:8: FutureWarning: The default value of observed=False is deprecated and will change to observed=True in a future version of pandas. Specify observed=False to silence this warning and retain the current behavior
  income_pivot = pd.pivot_table(
C:\Users\denis\AppData\Local\Temp\ipykernel_8432\3947398407.py:8: FutureWarning: The default value of observed=False is deprecated and will change to observed=True in a future version of pandas. Specify observed=False to silence this warning and retain the current behavior
  income_pivot = pd.pivot_table(


average_income  median_income
income_category marital_status                               
низкий          Divorced              17073.62        15068.0
                Married               21759.51        22574.0
                Single                22741.20        22694.5
                Together              21746.41        23661.0
                Widow                 27229.50        27229.5
средний         Alone                 43789.00        35860.0
                Divorced              50962.35        51074.0
                Married               49923.64        50442.0
                Single                48439.79        48503.5
                Together              51347.02        51518.0
                Widow                 50569.15        49218.0
высокий         Divorced              84939.06        81361.0
                Married               80804.91        79336.0
                Single                80170.97        79761.0
                Together              81252.33        77845.0
                Widow                 77468.43        79800.0

В категории низкого дохода наименьшие показатели у разведенных (Divorced), а наибольшие - у вдовцов (Widow)

В средней категории доходы достаточно равномерно распределены между разными семейными статусами

В высокой категории наблюдается значительный рост доходов, особенно у разведенных (средний доход 84.9 тыс.)

Семейное положение оказывает заметное влияние на уровень доходов, при этом разведенные демонстрируют наибольшую вариативность - от минимальных доходов в низкой категории до максимальных в высокой.

Задание 25. Создать столбец - срок с момента регистрации клиента в компании
(расчетный). Создать с помощью этого столбца столбец “Категория
клиента” (с помощью категоризации). Выделить минимум 3 категории
(новый, старый, средний), фильтрацию для уровня лет после регистрации
выбрать самостоятельно, аргументировать выбор. Создать столбец
“Категория дохода (Income)” (с помощью категоризации). Выделить
минимум 3 категории (высокий, низкий, средний), фильтрацию для уровня
дохода выбрать самостоятельно, аргументировать выбор. Выполнить
фильтрацию: выбрать записи только со средним и высоким доходом +
новых клиентов + топ 2 семейных статуса (Marital_Status) по среднему
пребыванию в компании + топ 3 типа образования по среднему доходу.
Создать сводные таблицы: средний и медианный Income по Education и
Marital_Status; средний доход по категории клиента и Education

1. **Расчет длительности клиентства**  
   - Преобразование даты начала клиентства в datetime объект
   - Вычисление длительности в годах с учетом високосных лет (деление на 365.25)

2. **Категоризация клиентов**  
   `pd.cut()` создает три категории:
   - новый: [0, 1) год
   - средний: [1, 3) года  
   - старый: [3, +∞) лет

3. **Категоризация дохода**  
   Создание трех income групп:
   - низкий: [0, 40000)
   - средний: [40000, 80000)  
   - высокий: [80000, +∞)

4. **Определение топ-групп**  
   - `top_marital`: 2 семейных статуса с наибольшей средней длительностью клиентства
   - `top_education`: 3 типа образования с наибольшим средним доходом

5. **Многоуровневая фильтрация**  
   Фильтр включает клиентов которые:
   - Имеют средний или высокий доход
   - Являются новыми клиентами (< 1 года)
   - Принадлежат к топ-2 семейных статусов
   - Имеют топ-3 типа образования

6. **Построение сводных таблиц**  
   - `pivot1`: Средний и медианный доход по образованию и семейному положению
   - `pivot2`: Средний доход по категории клиента и образованию

In [395]:
current_date = datetime.now()
df['customer_since_dt'] = pd.to_datetime(df['customer_since'])
df['membership_duration'] = (current_date - df['customer_since_dt']).dt.days / 365.25 

duration_bins = [0, 1, 3, np.inf]
duration_labels = ['новый', 'средний', 'старый']
df['customer_category'] = pd.cut(df['membership_duration'], bins=duration_bins, labels=duration_labels, right=False)

income_bins = [0, 40000, 80000, np.inf]
income_labels = ['низкий', 'средний', 'высокий']
df['income_category'] = pd.cut(df['income'], bins=income_bins, labels=income_labels, right=False)

# Определяем топ-2 семейных статуса по среднему пребыванию в компании
top_marital = df.groupby('marital_status')['membership_duration'].mean().nlargest(2).index.tolist()

# Определяем топ-3 типа образования по среднему доходу
top_education = df.groupby('education')['income'].mean().nlargest(3).index.tolist()

filtered_df = df[
    (df['income_category'].isin(['средний', 'высокий'])) &  
    (df['customer_category'] == 'новый') &  
    (df['marital_status'].isin(top_marital)) &  
    (df['education'].isin(top_education))  
]

pivot1 = pd.pivot_table(
    filtered_df,
    values='income',
    index=['education', 'marital_status'],
    aggfunc=['mean', 'median']
).round(2)

pivot2 = pd.pivot_table(
    filtered_df,
    values='income',
    index=['customer_category', 'education'],
    aggfunc='mean'
).round(2)

print("Топ-2 семейных статуса по среднему пребыванию:", top_marital)
print("Топ-3 типа образования по среднему доходу:", top_education)
print(f"Количество записей после фильтрации: {len(filtered_df)}")
print("\nСводная таблица 1: Средний и медианный доход по образованию и семейному положению")
display(pivot1)
print("\nСводная таблица 2: Средний доход по категории клиента и образованию")
display(pivot2)

C:\Users\denis\AppData\Local\Temp\ipykernel_8432\1811543351.py:14: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  top_marital = df.groupby('marital_status')['membership_duration'].mean().nlargest(2).index.tolist()


Топ-2 семейных статуса по среднему пребыванию: ['Divorced', 'Together']
Топ-3 типа образования по среднему доходу: ['PhD', 'Master', 'Graduation']
Количество записей после фильтрации: 0

Сводная таблица 1: Средний и медианный доход по образованию и семейному положению


C:\Users\denis\AppData\Local\Temp\ipykernel_8432\1811543351.py:17: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  top_education = df.groupby('education')['income'].mean().nlargest(3).index.tolist()
C:\Users\denis\AppData\Local\Temp\ipykernel_8432\1811543351.py:26: FutureWarning: The default value of observed=False is deprecated and will change to observed=True in a future version of pandas. Specify observed=False to silence this warning and retain the current behavior
  pivot1 = pd.pivot_table(
C:\Users\denis\AppData\Local\Temp\ipykernel_8432\1811543351.py:26: FutureWarning: The default value of observed=False is deprecated and will change to observed=True in a future version of pandas. Specify observed=False to silence this warning and retain the current behavior
  pivot1 = pd.pivot_table(
C:\User

Empty DataFrame
Columns: []
Index: []


Сводная таблица 2: Средний доход по категории клиента и образованию


Empty DataFrame
Columns: []
Index: []

После применения всех фильтров не осталось ни одной записи (0 записей), что указывает на слишком строгие условия отбора.